server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -454,21 +427,20 @@ public void close(Mode newMode) {
             assert closed == false : "CandidateJoinAccumulator closed";
             closed = true;
             if (newMode == Mode.LEADER) {
-                final List<JoinTaskExecutor.Task> pendingAsTasks = new ArrayList<>();
-                joinRequestAccumulator.forEach(
-                    (node, listener) -> pendingAsTasks.add(
-                        new JoinTaskExecutor.Task(node, joinReasonService.getJoinReason(node, Mode.CANDIDATE), listener)
-                    )
-                );
-
-                final String stateUpdateSource = "elected-as-master ([" + pendingAsTasks.size() + "] nodes joined)";
+                final JoinTask joinTask = new JoinTask(joinRequestAccumulator.entrySet().stream().map(entry -> {
+                    final DiscoveryNode discoveryNode = entry.getKey();
+                    final ActionListener<Void> listener = entry.getValue();
+                    return new JoinTask.NodeJoinTask(
+                        discoveryNode,
+                        joinReasonService.getJoinReason(discoveryNode, Mode.CANDIDATE),
+                        listener
+                    );
+                }).collect(Collectors.toList()), true);
```

Comment:
Please consider using `.toList()` instead of `.collect(Collectors.toList())` in new code. It is concise and also produces immutable list